services/blockstorage/block_sync.go

```diff
@@ -0,0 +1,323 @@
+			return
+		case event = <-b.Events:
+			continue
+		}
+	}
+}
+
+func (b *BlockSync) dispatchEvent(state blockSyncState, event interface{}, blockAvailabilityResponses []*gossipmessages.BlockAvailabilityResponseMessage) (blockSyncState, []*gossipmessages.BlockAvailabilityResponseMessage) {
+	switch event.(type) {
+	case *gossipmessages.BlockAvailabilityRequestMessage:
+		message := event.(*gossipmessages.BlockAvailabilityRequestMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.SourceHandleBlockAvailabilityRequest(message)
+		}
+	case *gossipmessages.BlockAvailabilityResponseMessage:
+		message := event.(*gossipmessages.BlockAvailabilityResponseMessage)
+
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			err := b.PetitionerHandleBlockAvailabilityResponse(message)
+
+			if err != nil {
+				b.reporting.Info("received bad block availability response", log.Error(err))
+			} else {
+				blockAvailabilityResponses = append(blockAvailabilityResponses, message)
+			}
+		}
+	case *gossipmessages.BlockSyncRequestMessage:
+		message := event.(*gossipmessages.BlockSyncRequestMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.SourceHandleBlockSyncRequest(message)
+			return BLOCK_SYNC_STATE_IDLE, blockAvailabilityResponses
+		}
+	case *gossipmessages.BlockSyncResponseMessage:
+		message := event.(*gossipmessages.BlockSyncResponseMessage)
+		if fromMe := message.Sender.SenderPublicKey().Equal(b.config.NodePublicKey()); !fromMe {
+			b.PetitionerHandleBlockSyncResponse(message)
+			return BLOCK_SYNC_STATE_IDLE, blockAvailabilityResponses
+		}
+	}
+
+	return state, blockAvailabilityResponses
+}
+
+func (b *BlockSync) PetitionerBroadcastBlockAvailabilityRequest() error {
+	lastCommittedBlockHeight := b.storage.LastCommittedBlockHeight()
+	firstBlockHeight := lastCommittedBlockHeight + 1
+	lastBlockHeight := b.storage.LastCommittedBlockHeight() + primitives.BlockHeight(b.config.BlockSyncBatchSize())
+
+	b.reporting.Info("broadcast block availability request",
+		log.Stringable("first-block-height", firstBlockHeight),
+		log.Stringable("last-block-height", lastBlockHeight))
+
+	input := &gossiptopics.BlockAvailabilityRequestInput{
+		Message: &gossipmessages.BlockAvailabilityRequestMessage{
+			Sender: (&gossipmessages.SenderSignatureBuilder{
+				SenderPublicKey: b.config.NodePublicKey(),
+			}).Build(),
+			SignedBatchRange: (&gossipmessages.BlockSyncRangeBuilder{
+				BlockType:                gossipmessages.BLOCK_TYPE_BLOCK_PAIR,
+				LastBlockHeight:          lastBlockHeight,
+				FirstBlockHeight:         firstBlockHeight,
+				LastCommittedBlockHeight: lastCommittedBlockHeight,
+			}).Build(),
+		},
+	}
+
+	_, err := b.gossip.BroadcastBlockAvailabilityRequest(input)
+	return err
+}
+
+func (b *BlockSync) PetitionerHandleBlockSyncResponse(message *gossipmessages.BlockSyncResponseMessage) {
```

Comment:
This method should be private, if it needs to be public for the tests it means that this is a smell and these tests should be converted to unit tests. If it's too complicated to change now, add a TODO and we'll do it once we introduce unit tests

Reply:
Fixed